File: packages/uc04/src/uc04/core/text.py
"""Deterministic text primitives shared by matching, tagging and fingerprinting.

Ported from the TypeScript reference. Two behaviours were re-verified rather than assumed,
because they do not translate byte-for-byte:

* ``\\b`` word-boundary semantics differ between the JS and Python regex engines around
  apostrophes and digits, so every pattern that relies on them is tested individually
  (see tests/test_regex_verification.py).
* The stemmer is deliberately conservative - no ``er``/``ed`` rules, which would turn
  "answer" into "answ" and silently change every downstream score.
"""

from __future__ import annotations

import re
from collections.abc import Iterable, Sequence

STOPWORDS: frozenset[str] = frozenset(
    """
    a an the and or but if then than that this these those is are was were be been being am
    do does did doing have has had having i you he she it we they me him her us them my your
    his its our their to of in on at for with about as by from into over under again further
    so can could would should will shall may might must not no nor very just also too there
    here what which who whom when where why how please tell explain give some any all more
    most other such own same because
    """.split()
)

_PUNCT = re.compile(r"[^a-z0-9'\s-]")
#: Hyphens are split, not kept. "out-of-court" and "out of court" are the same phrase, and
#: treating them as different tokens made paraphrase detection and retrieval disagree with
#: themselves depending on how the author happened to punctuate.
_HYPHEN = re.compile(r"-+")
_WS = re.compile(r"\s+")
_SMART_QUOTES = re.compile("[‘’“”]")

#: Sentence splitter. Python's re supports fixed-width lookbehind, which covers this pattern.
_SENTENCE_SPLIT = re.compile(r"(?<=[.!?])\s+")

#: Conservative suffix rules, longest first.
_SUFFIXES: tuple[str, ...] = ("ations", "ation", "ings", "ing", "ies", "es", "s")


def normalize_text(value: str) -> str:
    lowered = _SMART_QUOTES.sub("'", value.lower())
    unhyphenated = _HYPHEN.sub(" ", lowered)
    return _WS.sub(" ", _PUNCT.sub(" ", unhyphenated)).strip()


def tokenize(value: str) -> list[str]:
    normalized = normalize_text(value)
    return [token for token in normalized.split(" ") if token] if normalized else []


def stem(token: str) -> str:
    for suffix in _SUFFIXES:
        if len(token) > len(suffix) + 2 and token.endswith(suffix):
            trimmed = token[: -len(suffix)]
            return trimmed + "y" if suffix == "ies" else trimmed
    return token


def content_tokens(value: str) -> list[str]:
    return [stem(t) for t in tokenize(value) if len(t) > 2 and t not in STOPWORDS]


def unique_tokens(value: str) -> list[str]:
    seen: dict[str, None] = {}
    for token in content_tokens(value):
        seen.setdefault(token, None)
    return list(seen)


def jaccard(left: Sequence[str], right: Sequence[str]) -> float:
    if not left and not right:
        return 1.0
    a, b = set(left), set(right)
    intersection = len(a & b)
    union = len(a) + len(b) - intersection
    return intersection / union if union else 0.0


def stable_hash(value: str) -> str:
    """FNV-1a 64-bit, hex. Stable across processes and runtimes."""
    h = 0xCBF29CE484222325
    for char in value:
        h ^= ord(char)
        h = (h * 0x100000001B3) & 0xFFFFFFFFFFFFFFFF
    return format(h, "016x")


def sentences(value: str) -> list[str]:
    return [s.strip() for s in _SENTENCE_SPLIT.split(value) if s.strip()]


def truncate_words(value: str, max_words: int) -> str:
    words = value.split()
    if len(words) <= max_words:
        return value.strip()
    return " ".join(words[:max_words]) + "..."


def contains_all(haystack: Iterable[str], needles: Sequence[str]) -> bool:
    pool = set(haystack)
    return bool(needles) and all(n in pool for n in needles)

File: reference/uc04-typescript/src/core/text.ts
/** Deterministic text utilities shared by retrieval, fingerprinting and the classifier. */

const STOPWORDS = new Set([
  'a','an','the','and','or','but','if','then','than','that','this','these','those','is','are','was','were','be','been',
  'being','am','do','does','did','doing','have','has','had','having','i','you','he','she','it','we','they','me','him',
  'her','us','them','my','your','his','its','our','their','to','of','in','on','at','for','with','about','as','by','from',
  'into','over','under','again','further','so','can','could','would','should','will','shall','may','might','must','not',
  'no','nor','very','just','also','too','there','here','what','which','who','whom','when','where','why','how','please',
  'tell','explain','give','me','some','any','all','more','most','other','such','own','same','because',
]);

export function normalizeText(input: string): string {
  return input
    .toLowerCase()
    .replace(/[‘’“”]/g, "'")
    .replace(/[^a-z0-9'\s-]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

/** Lowercased, punctuation-stripped tokens (stopwords kept). */
export function tokenize(input: string): string[] {
  const normalized = normalizeText(input);
  if (!normalized) return [];
  return normalized.split(' ').filter(Boolean);
}

/** Content tokens: stopwords removed, very short tokens dropped, light singularisation. */
export function contentTokens(input: string): string[] {
  return tokenize(input)
    .filter((t) => t.length > 2 && !STOPWORDS.has(t))
    .map(stem);
}

/** Extremely small, deterministic suffix stemmer - enough for lexical overlap scoring. */
export function stem(token: string): string {
  let t = token;
  // Deliberately conservative: no 'er'/'ed' rules, which would mangle words like "answer".
  for (const suffix of ['ations', 'ation', 'ings', 'ing', 'ies', 'es', 's']) {
    if (t.length > suffix.length + 2 && t.endsWith(suffix)) {
      t = t.slice(0, -suffix.length);
      if (suffix === 'ies') t += 'y';
      break;
    }
  }
  return t;
}

export function uniqueTokens(input: string): string[] {
  return Array.from(new Set(contentTokens(input)));
}

export function jaccard(a: readonly string[], b: readonly string[]): number {
  if (a.length === 0 && b.length === 0) return 1;
  const setA = new Set(a);
  const setB = new Set(b);
  let intersection = 0;
  for (const t of setA) if (setB.has(t)) intersection += 1;
  const union = setA.size + setB.size - intersection;
  return union === 0 ? 0 : intersection / union;
}

/** Overlap of `query` tokens covered by `target` tokens, weighted toward rarer long tokens. */
export function coverageScore(queryTokens: readonly string[], targetTokens: readonly string[]): number {
  if (queryTokens.length === 0) return 0;
  const target = new Set(targetTokens);
  let matchedWeight = 0;
  let totalWeight = 0;
  for (const token of new Set(queryTokens)) {
    const weight = 1 + Math.min(token.length, 12) / 12;
    totalWeight += weight;
    if (target.has(token)) matchedWeight += weight;
  }
  return totalWeight === 0 ? 0 : matchedWeight / totalWeight;
}

/** FNV-1a 64-bit (as hex) - stable across processes, no crypto dependency needed. */
export function stableHash(input: string): string {
  let hash = 0xcbf29ce484222325n;
  const prime = 0x100000001b3n;
  const mask = 0xffffffffffffffffn;
  for (let i = 0; i < input.length; i += 1) {
    hash ^= BigInt(input.charCodeAt(i));
    hash = (hash * prime) & mask;
  }
  return hash.toString(16).padStart(16, '0');
}

/** Split prose into sentences, preserving order. */
export function sentences(input: string): string[] {
  return input
    .split(/(?<=[.!?])\s+/)
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
}

export function truncateWords(input: string, maxWords: number): string {
  const words = input.split(/\s+/).filter(Boolean);
  if (words.length <= maxWords) return input.trim();
  return `${words.slice(0, maxWords).join(' ')}...`;
}
